### PR TITLE
feat(wmg): 3991 SVG download with MutationObserver

### PR DESCRIPTION
**Pull request by tihuan**
## Reason for Change

- #3991 
- If the reason for this PR's code changes are not clear in the issue, state value/impact

1. Interactive heatmap is in canvas mode, which is NOT suitable for downloads, so we need to re-render heatmap in SVG before triggering the download. However, if we trigger download before the heatmap is done rendering, we'll miss elements in the downloaded files, thus the need to add MutationObserver to ensure the heatmap is done rendering before triggering download

## Changes

1. Add MutationObserver to check if the heatmap is done rendering SVG elements before downloading images
2. Move SVG render functions into `util.ts`
3. Re-enable canvas mode for interactive heatmap
4. Minor TS fixes to avoid using `!`

## Testing steps

1. Go to GE
2. In Genes selector, paste and enter the following to populate many genes:
```
MALAT1,TSPAN6,TNMD,DPM1,SCYL3,C1orf112,FGR,CFH,FUCA2,GCLC,NFYA,STPG1,NIPAL3,LAS1L,ENPP4,SEMA3F,CFTR,ANKIB1,CYP51A1,KRIT1,RAD52,MYH16,BAD,LAP3,CD99,HS3ST1,AOC1,MALAT1,WNT16,HECW1,MAD1L1,LASP1,SNX11,TMEM176A,M6PR,KLHL13,CYP26B1,ICA1,DBNDD1,ALS2,CASP10,CFLAR,TFPI,NDUFAF7,RBM5,MTMR7,SLC7A2,ARF5,SARM1,POLDIP2,PLXND1,AK2,CD38,FKBP4,RBM6,KDM1A,CAMKK1,RECQL,VPS50,HSPB6,ARHGAP33,NDUFAB1,PDK4,SLC22A16,ZMYND10
```
3. Add a few tissues
4. Once rendering is done, trigger download
5. Check the downloaded files and make sure they match the fully rendered heatmap in the browser

## Notes for Reviewer

Thank you!

### CODE REVIEW

File: frontend/src/views/WheresMyGene/components/GeneSearchBar/components/SaveImage/index.tsx [hunk 1]
```diff
@@ -328,375 +273,115 @@ function renderDots({
   return chart;
 }
 
-const NAME_SPACE_URI = "http://www.w3.org/2000/svg";
```
Comment (tihuan) [inline frontend/src/views/WheresMyGene/components/GeneSearchBar/components/SaveImage/index.tsx:331]: Here and below, moved SVG rendering functions into `/utils.ts`

Net change in this file is adding MutationObserver and wrap `download` function in a debounce

File: frontend/src/views/WheresMyGene/components/GeneSearchBar/components/SaveImage/index.tsx [hunk 2]
```diff
@@ -89,100 +94,41 @@ export default function SaveImage({
...
+    }, MUTATION_OBSERVER_TIMEOUT);
+
+    const observer = new MutationObserver(callback);
+
+    /**
+     * (thuang): We only want to call download() once, AFTER the heatmap is fully rendered.
+     * We assume the heatmap is fully rendered after no more changes to the DOM tree
+     * after `MUTATION_OBSERVER_TIMEOUT` milliseconds
+     */
+    const download = debounce(
```
Comment (tihuan) [inline frontend/src/views/WheresMyGene/components/GeneSearchBar/components/SaveImage/index.tsx:115]: debounce the MutationObserver callback function, so we only trigger download once after the DOM stops mutation

File: frontend/src/views/WheresMyGene/components/GeneSearchBar/components/SaveImage/utils.ts [hunk 3]
```diff
@@ -0,0 +1,380 @@
+import { Tissue } from "src/views/WheresMyGene/common/types";
```
Comment (tihuan) [inline frontend/src/views/WheresMyGene/components/GeneSearchBar/components/SaveImage/utils.ts:1]: this file has all the SVG rendering functions. No business logics changed here beside a few tiny TypeScript changes to remove the usage of assertion `!`

File: frontend/src/views/WheresMyGene/components/HeatMap/components/Chart/index.tsx [hunk 4]
```diff
@@ -119,8 +119,7 @@ export default memo(function Chart({
     setIsChartInitialized(true);
 
     const chart = init(current, EMPTY_OBJECT, {
-      renderer: "svg",
-      // renderer: echartsRendererMode,
+      renderer: echartsRendererMode,
```
Comment (tihuan) [inline frontend/src/views/WheresMyGene/components/HeatMap/components/Chart/index.tsx:122]: this enables svg vs. canvas mode again

File: frontend/src/views/WheresMyGene/components/HeatMap/components/Chart/index.tsx [hunk 5]
```diff
@@ -98,7 +98,7 @@ export default memo(function Chart({
       const pointInGrid = chart.convertFromPixel("grid", [offsetX, offsetY]);
 
       setCursorOffset([pageX, pageY]);
-      setCurrentIndices(pointInGrid);
+      if (pointInGrid) setCurrentIndices(pointInGrid);
```
Comment (tihuan) [inline frontend/src/views/WheresMyGene/components/HeatMap/components/Chart/index.tsx:101]: guard to not set indices when unavailable

**Review by ashin-czi — APPROVED** [review]
lgtm!

**Pull request merged**